### ISSUE TITLE
Different test runners show different results for NUnit.Framework.TestContextOneTimeTearDownTests

### ISSUE THREAD

**Issue by YuriyOborozhnyi** [nunit/nunit#2906]
Accidentally I've found that `TestContextOneTimeTearDownTests` test fixture behaves differently depending on which test runner is used. Here below I am referencing to the following source code, especially `OneTimeTearDown` method.
https://github.com/nunit/nunit/blob/5c4394af96570afb559f7bce238aa859b9ad0c61/src/NUnitFramework/tests/TestContextTests.cs#L452-L486

# Steps To Reproduce

1. If you are running tests from IDE, set a breakpoint 🔴 in `OneTimeTearDown` method.
2. Run `TestContextOneTimeTearDownTests` test fixture as a whole in `nunit.framework` test assembly, but not each test separately.

# Actual Result
## ReSharper
_The version of ReSharper is 2018.1.2._
**❌ `TestContextOneTimeTearDownTests` failed.**
![image](https://user-images.githubusercontent.com/22633878/41839375-c57ef6e2-786b-11e8-83bf-82e32d6e2d8c.png)
- `context.Result.PassCount` = 2
- `context.Result.SkipCount` = 0 _(the reason of failure)_

## Visual Studio
**✔️ Tests are passed.**
![image](https://user-images.githubusercontent.com/22633878/41841297-96d40e18-7870-11e8-83ba-ec95343d2800.png)
- `context.Result.PassCount` = 3
- `context.Result.SkipCount` = 0

`Assert.That(context.Result.PassCount, Is.EqualTo(2));` actually throws `AssertionException`, but seems it isn't counted by VS, because all is green.

## NUnit Console
**✔️ Tests are passed.**
![image](https://user-images.githubusercontent.com/22633878/41841462-fe5862e6-7870-11e8-8307-e9d0cd8098bb.png)
- Passed: 2
- Skipped: 1

# Expected Result
Since all 3 test runners provide different results, I am not sure what is correct. It depends on how "Explicit" tests and "TearDown" failures should be treated.

**Comment by YuriyOborozhnyi** [comment]
Guys, seems this issue is about not a single issue actually, but I am not sure. I can create separate issues when I understand better what's going on.

As I can understand, https://github.com/nunit/nunit/issues/219 was aimed to tackle "TearDown" failures, but I still don't understand how they should be processed.

Also there is https://github.com/nunit/nunit/issues/937 which is supposed to deal with "Explicit" tests, but I still don't understand how they should be processed now.

**Comment by YuriyOborozhnyi** [comment]
@jnm2, please take a look. I am pretty new into your repository, so don't know if I should attach any label to this issue.

**Comment by CharliePoole** [comment]
The test fixture tests what your TearDown code is able to see in various circumstances.

The explicit test is expected to be skipped unless you run it directly. However, not all runners work correctly in this respect. Obviously, the behavior of the runner is outside the control of the framework tests.

The console runner behaves correctly by not running that test. It counts the test as skipped because of being explicit.

Resharper runs the test correctly as well... that is, it doesn't run it but doesn't count it or report it. This is how NUnit previously handled Explicit in V2. Both reporting approaches have their fans.

The test explorer does not pass enough information to the adapter for us to determine how the test was selected in the gui. In fact, it passes in the I'd of every test in the fixture. So we execute all of them. This is a known problem and can't be fixed from the NUnit side.

I'd say there is no framework bug here. The framework is being told to do different things by the runners and the runners are handling the result differently.

**Comment by YuriyOborozhnyi** [comment]
@CharliePoole, thank you for the detailed explanation!

I want to ask the last one: as I wrote, `Assert.That(context.Result.PassCount, Is.EqualTo(2))` fails in `OneTimeTearDown` when tests are run from Visual Studio. But finally all tests are green. Is it a bug or not? Looks similar to https://github.com/nunit/nunit/issues/219.

**Comment by CharliePoole** [comment]
Look back at the paragraph I wrote about test explorer. To put it another way, Visual Studio runs all three tests when it should only run one of them.

**Comment by CharliePoole** [comment]
It could be argued that this is s bad unit test, since an external factor can make it fail. Of course the team in the past never ran tests in visual studio.

**Comment by YuriyOborozhnyi** [comment]
@CharliePoole, got it. So basically any exception thrown in TearDown will be hidden if tests are run from Visual Studio...

**Comment by jnm2** [comment]
> It could be argued that this is s bad unit test, since an external factor can make it fail. Of course the team in the past never ran tests in visual studio.

I lean towards this perspective. If we remove the explicit test, what guarantee are we losing? If there is one, we can move it to the testdata assembly so that we control the explicit selection logic.

**Comment by jnm2** [comment]
@YuriyOborozhnyi Only members can label issues, so you're good. :D

**Comment by CharliePoole** [comment]
@jnm2 I agree. IMO, that's where the whole fixture should be. Also, the first two tests are the same from the perspective of what we are testing: both of them succeed. We need a failing test and a way to assert on what is seen by TearDown. No big priority however.

**Comment by jnm2** [comment]
@YuriyOborozhnyi, is this something you'd like to fix?

I'd suggest taking the same approach as https://github.com/nunit/nunit/blob/master/src/NUnitFramework/tests/Internal/SetUpTearDownTests.cs.

You'd move `TestContextOneTimeTearDownTests` to the testdata assembly. Then rather than asserting in its `OneTimeTearDown` as it's doing now, you'd introduce `Test` and `Result` properties on the fixture instance and set them from within `OneTimeTearDown` from of off `TestContext.CurrentContext`.

Then the test code in nunit.framework.tests would be similar to this:

https://github.com/nunit/nunit/blob/5c4394af96570afb559f7bce238aa859b9ad0c61/src/NUnitFramework/tests/Internal/SetUpTearDownTests.cs#L46-L54

But you'd be making assertions on `fixture.Test` and `fixture.Result`.
